crates/bevy_render/src/batching/mod.rs

```diff
@@ -73,24 +74,57 @@ pub trait GetBatchData {
     fn get_batch_data(
         param: &SystemParamItem<Self::Param>,
         query_item: &QueryItem<Self::Data>,
-    ) -> (Self::BufferData, Option<Self::CompareData>);
+    ) -> Option<(Self::BufferData, Option<Self::CompareData>)>;
 }
 
 /// Batch the items in a render phase. This means comparing metadata needed to draw each phase item
 /// and trying to combine the draws into a batch.
-pub fn batch_and_prepare_render_phase<I: CachedRenderPipelinePhaseItem, F: GetBatchData>(
+pub fn batch_and_prepare_render_phase<
+    I: CachedRenderPipelinePhaseItem,
+    F: GetBatchData<Data = Entity>,
+>(
+    gpu_array_buffer: ResMut<GpuArrayBuffer<F::BufferData>>,
+    mut views: Query<&mut RenderPhase<I>>,
+    param: StaticSystemParam<F::Param>,
+) {
+    let gpu_array_buffer = gpu_array_buffer.into_inner();
+    let system_param_item = param.into_inner();
+
+    let process_item = |item: &mut I| {
+        let (buffer_data, compare_data) = F::get_batch_data(&system_param_item, &item.entity())?;
+        let buffer_index = gpu_array_buffer.push(buffer_data);
+
+        let index = buffer_index.index.get();
+        *item.batch_range_mut() = index..index + 1;
+        *item.dynamic_offset_mut() = buffer_index.dynamic_offset;
+
+        if I::AUTOMATIC_BATCHING {
+            compare_data.map(|compare_data| BatchMeta::new(item, compare_data))
+        } else {
+            None
+        }
+    };
+
+    apply_batch::<I, F>(&mut views, process_item);
+}
+
+/// Batch the items in a render phase with Query to fetch its required data. If Query is unnecessary ,using [`batch_and_prepare_render_phase`] will be more efficient, Not used yet
+pub fn batch_and_prepare_render_phase_with_query<
```

Comment:
If it's not being used, it's best to remove it entirely, as well as the associated type. Unless @superdump or @robtfm have any other objections to dropping this functionality from the trait.

Reply:
yeah, I also intend to remove it , IMO,we should not do any related-query during batch&prepare phase,not only does this incur additional overhead, but it also hinders the potential for parallelization in the future. once Bevy has a more powerful parallel infrastructure, we can easily parallelize this phase.

Reply:
ok so currently `F::Filter` gets checked by `query.get()` and is effectively checked again by `mesh_instances.get()`, so the cost of the system reducing by 50% when all items match makes sense (and there should still be a smaller benefit if not all phase items match the query, tending to equal as the number of phase item types tends to infinity).

the query was because originally we stored data on the entity instead of in the non-ecs-based `RenderMeshInstances`. i'm not sure we are fully committed to that approach, but the change wouldn't be hard to revert if we need to and the impact is worthwhile. any third party users can follow the same pattern as well. let's nuke the with_query variant.

Reply:
IIRC,`F::Filter ` or `mesh_instances.get()` overhead is relatively small , the bottleneck lies in performing query.get(entity) for each entity,it needs to introduce  extra `init_fetch` and `set_archetype/table` per entity, therefore, when none of the phase items match, the performance gain is still quite significant.